go-controller/pkg/ovn/topology_version.go

```diff
@@ -0,0 +1,148 @@
+	if ver <= ovntypes.OvnPortBindingTopoVersion {
+		err = addressset.NonDualStackAddressSetCleanup(oc.nbClient)
+	}
+	return err
+}
+
+// reportTopologyVersion saves the topology version to two places:
+// - an ExternalID on the ovn_cluster_router LogicalRouter in nbdb
+// - a ConfigMap. This is used by nodes to determine the cluster's topology
+func (oc *Controller) reportTopologyVersion(ctx context.Context) error {
+	currentTopologyVersion := strconv.Itoa(ovntypes.OvnCurrentTopologyVersion)
+	logicalRouterRes := []nbdb.LogicalRouter{}
+	ctx, cancel := context.WithTimeout(ctx, ovntypes.OVSDBTimeout)
+	defer cancel()
+	if err := oc.nbClient.WhereCache(func(lr *nbdb.LogicalRouter) bool {
+		return lr.Name == ovntypes.OVNClusterRouter
+	}).List(ctx, &logicalRouterRes); err != nil {
+		return fmt.Errorf("failed in retrieving %s, error: %v", ovntypes.OVNClusterRouter, err)
+	}
+	// Update topology version on distributed cluster router
+	logicalRouterRes[0].ExternalIDs["k8s-ovn-topo-version"] = currentTopologyVersion
+	logicalRouter := nbdb.LogicalRouter{
+		Name:        ovntypes.OVNClusterRouter,
+		ExternalIDs: logicalRouterRes[0].ExternalIDs,
+	}
+	opModel := libovsdbops.OperationModel{
+		Name:           logicalRouter.Name,
+		Model:          &logicalRouter,
+		ModelPredicate: func(lr *nbdb.LogicalRouter) bool { return lr.Name == ovntypes.OVNClusterRouter },
+		OnModelUpdates: []interface{}{
+			&logicalRouter.ExternalIDs,
+		},
+		ErrNotFound: true,
+	}
+	if _, err := oc.modelClient.CreateOrUpdate(opModel); err != nil {
+		return fmt.Errorf("failed to generate set topology version in OVN, err: %v", err)
+	}
+	klog.Infof("Updated Logical_Router %s topology version to %s", ovntypes.OVNClusterRouter, currentTopologyVersion)
+
+	// Report topology version in a ConfigMap
+	// (we used to report this via annotations on our Node)
+	cm := corev1apply.ConfigMap(ovntypes.OvnK8sStatusCMName, globalconfig.Kubernetes.OVNConfigNamespace)
+	cm.Data = map[string]string{ovntypes.OvnK8sStatusKeyTopoVersion: currentTopologyVersion}
+	if _, err := oc.client.CoreV1().ConfigMaps(globalconfig.Kubernetes.OVNConfigNamespace).Apply(ctx, cm, metav1.ApplyOptions{
+		Force:        true,
+		FieldManager: "ovn-kubernetes",
+	}); err != nil {
+		return err
+	}
+
+	klog.Infof("Updated ConfigMap %s/%s topology version to %s", *cm.Namespace, *cm.Name, currentTopologyVersion)
+
+	return oc.cleanTopologyAnnotation()
+}
+
+// Remove the old topology annotation from nodes, if it exists.
+func (oc *Controller) cleanTopologyAnnotation() error {
+	// Unset the old topology annotation on all Node objects
+	nodes, err := oc.watchFactory.GetNodes()
+	if err != nil {
+		return err
+	}
+	anno := ovntypes.OvnK8sTopoAnno //nolint // otherwise we get deprecation warnings (this variable is deprecated)
+	for _, node := range nodes {
+		if _, ok := node.Annotations[anno]; !ok {
+			continue
+		}
+		err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
```

Comment:
nit: in this PR there is an explicit notion of passing context everywhere to try to cancel things if ovnkube node shuts down. Looks like this function does not respect a context cancel. maybe a todo later

Reply:
Right, the assumption is that the cancelled context would cause the initial Get to fail, but because our kube wrapper drops contexts... it doesn't quite work.